### PR TITLE
[processor/servicegraph]add virtual node feature.

### PR DESCRIPTION
**Pull request by JaredTan95**
Signed-off-by: Jared Tan <jian.tan@daocloud.io>

**Description:** 
impl https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/17196 

**Link to tracking Issue:** <Issue number if applicable>

**Testing:** <Describe what testing was performed and which tests were added.>

**Documentation:** <Describe the documentation added.>

### CODE REVIEW

**Comment by JaredTan95** [comment]
label -stale

File: processor/servicegraphprocessor/internal/store/store_test.go
Comment (jpkrohling) [inline processor/servicegraphprocessor/internal/store/store_test.go:139]: I believe a new test is deserved here, to exercise the new conditions. Can you add a test that will yield true when  trySpeculateEvictHead is called, and one which will yield false?

**Review by jpkrohling — COMMENTED** [review]
@mapno, would you be able to review this one as well?

File: processor/servicegraphprocessor/internal/store/store_test.go
Comment (JaredTan95) [reply processor/servicegraphprocessor/internal/store/store_test.go:139]: 👌

**Review by mapno — COMMENTED** [review]
I like the idea, but I'd rather keep it from leaking into `internal/store` if possible. We can have it be part of [`p.onExpire`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/servicegraphprocessor/processor.go#L283-L292), which is a practically unused function at the moment. Ie. when the edge expires, we check if it has peer attributes, and we aggregate the metrics if that's the case.

Also, this should be a configurable feature. Mainly b/c there aren't official semantic conventions that identify the type of requests we're trying to support.

**Comment by JaredTan95** [comment]
> I like the idea, but I'd rather keep it from leaking into `internal/store` if possible. We can have it be part of [`p.onExpire`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/servicegraphprocessor/processor.go#L283-L292), which is a practically unused function at the moment. Ie. when the edge expires, we check if it has peer attributes, and we aggregate the metrics if that's the case.
> 
> Also, this should be a configurable feature. Mainly b/c there aren't official semantic conventions that identify the type of requests we're trying to support.

@jpkrohling @mapno thx for your review, I polished those logical

**Comment by JaredTan95** [comment]
But a little confused, `VerifyXXX` in the test is not guaranteed to be called back every time.

File: processor/servicegraphprocessor/processor.go
Comment (mapno) [inline processor/servicegraphprocessor/processor.go:494]: This function is not really needed. It is being called in `OnExpire`, which is called by the store when an edge has expired. A lot of the code here is redundant. It can be simplified to just checking the peer labels and aggregating the edge to the metrics.

**Review by mapno — COMMENTED** [review]
Thanks for the PR. I think we're close, it needs to be simplified a bit. Also, docs in the README about this flag would be great.

**Comment by JaredTan95** [comment]
lint ci time out : https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/4122369157/jobs/7119166768#step:9:43

File: processor/servicegraphprocessor/internal/store/edge.go
Comment (mapno) [inline processor/servicegraphprocessor/internal/store/edge.go:65]: We can now unexport these methods

File: processor/servicegraphprocessor/processor.go
Comment (mapno) [inline processor/servicegraphprocessor/processor.go:315]: `OnComplete` is called when an edge is completed after being updated. Expired edges can't be complete, so we can remove this call.

File: processor/servicegraphprocessor/processor.go
Comment (mapno) [inline processor/servicegraphprocessor/processor.go:303]: This can be cleaned up a bit. The log above can be confusing if we end up recording the metric, maybe move it down to when the edge is expired and not aggregated.

File: processor/servicegraphprocessor/processor.go
Comment (mapno) [inline processor/servicegraphprocessor/processor.go:51]: Nitpick: WDYT of a more descriptive name like:

```suggestion
	PeerAttributes = []string{semconv.AttributeDBName, semconv.AttributeNetSockPeerAddr, semconv.AttributeNetPeerName, semconv.AttributeRPCService, semconv.AttributeHTTPURL, semconv.AttributeHTTPTarget}
```

File: processor/servicegraphprocessor/README.md
Comment (mapno) [inline processor/servicegraphprocessor/README.md:17]: Can you add what attributes are used to get the server name?

File: processor/servicegraphprocessor/processor.go
Comment (mapno) [inline processor/servicegraphprocessor/processor.go:511]: Nitpick: I think this is more idiomatic.

```suggestion
		if peer, ok := peers[s]; ok {
			peerStr = peer
```

File: processor/servicegraphprocessor/processor.go
Comment (mapno) [inline processor/servicegraphprocessor/processor.go:278]: If we're only using the first find of the list, we can break here and save some cycles per edge. `findAttributeValue` is a brute force search.

File: processor/servicegraphprocessor/processor.go
Comment (mapno) [inline processor/servicegraphprocessor/processor.go:214]: Only call when `VirtualNodeFeatureEnabled = true`?

File: processor/servicegraphprocessor/processor.go
Comment (mapno) [inline processor/servicegraphprocessor/processor.go:306]: I'm a bit worried about this one. Basically, if the feature is enabled, any trace with a server span and no client span will end up marked as `virtual_node`, when it may not be the case. The client span could just be missing, or arriving late.

We could add some logic to check if the server span is an orphan—ie. it doesn't have a parent/client span. In that case, it's fair to assume the client is from _the outside_. What do you think?

**Review by mapno — COMMENTED** [review]
I think this is making great progress. Thanks for the patience and for addressing my comments. Left a few more, many of them just of style.

File: processor/servicegraphprocessor/processor.go
Comment (JaredTan95) [reply processor/servicegraphprocessor/processor.go]: yes, I forgot this.

File: processor/servicegraphprocessor/processor.go
Comment (JaredTan95) [reply processor/servicegraphprocessor/processor.go:313]: You're right, but I think there will be more complex processing logic involved, and we can leave it to the next PR or implement it after more thought.

**Comment by JaredTan95** [comment]
> I think this is making great progress. Thanks for the patience and for addressing my comments. Left a few more, many of them just of style.

thx for your review, you helped me a lot something I didn't noticed.

File: processor/servicegraphprocessor/processor.go
Comment (mapno) [reply processor/servicegraphprocessor/processor.go:313]: Alright, that's fair. I'll open an issue to track this work. In the meantime, I'd add a TODO or comment to be explicit.

File: processor/servicegraphprocessor/processor.go
Comment (mapno) [inline processor/servicegraphprocessor/processor.go:322]: The processor's `onComplete` and the store's `OnComplete` are essentially the same function. We can use the processor's here and unexport `OnComplete` and `OnExpire` in the store.

**Review by mapno — APPROVED** [review]
Alright, just left two minor comments. Otherwise, this is good to go for me. Thanks for the work! I'll leave it to OTel maintainers to final approve and merge.

File: processor/servicegraphprocessor/processor.go
Comment (mapno) [reply processor/servicegraphprocessor/processor.go:322]: That's a good point. Another alternative is simply calling `p.aggregateMetricsForEdge(e)`.

File: processor/servicegraphprocessor/config.go
Comment (jpkrohling) [inline processor/servicegraphprocessor/config.go:42]: I'm a bit outdated, but is this how we do feature flags within components? Perhaps @djaglowski knows?

**Review by jpkrohling — COMMENTED** [review]
Other than the question about the feature flag, this looks good to me.

File: processor/servicegraphprocessor/config.go
Comment (djaglowski) [reply processor/servicegraphprocessor/config.go:42]: Check out the [featuregate](https://github.com/open-telemetry/opentelemetry-collector/tree/main/featuregate) package. It has a good example in the readme, but basically:
1. Use an init func to register a global gate. This returns the value of the gate according to step 2.
2. The gate can only be managed at startup time via use of the `--feature-gates` command line arg. 
3. Save the gate on the component and reference it as necessary.

Comment (JaredTan95) [reply processor/servicegraphprocessor/config.go:42]: thx, I will polish it with feature flags.

File: processor/servicegraphprocessor/processor_test.go
Comment (jpkrohling) [inline processor/servicegraphprocessor/processor_test.go:107]: Should you unset this when the test finishes?

Comment (JaredTan95) [reply processor/servicegraphprocessor/processor_test.go:107]: It doesn't matter if we don't reset it, but I'll reset it anyway

**Comment by JaredTan95** [comment]
we need to re-trigger the CI ?https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/4160007985/jobs/7196621241#step:15:337

**Comment by djaglowski** [comment]
Please rebase. The failing test was removed in #18519.

**Comment by JaredTan95** [comment]
> Please rebase. The failing test was removed in #18519.

ok

**Review by djaglowski — APPROVED** [review]

**Pull request merged**
